### PR TITLE
Regular expression support in vmod-cookie

### PR DESCRIPTION
**Pull request by lkarsten**
Hi.

This PR adds regular expression support in vmod-cookie. 

Please consider this for merging into varnish-modules.

### CODE REVIEW

File: src/vmod_cookie.c [hunk 1]
```diff
@@ -213,6 +220,58 @@ vmod_get(VRT_CTX, struct vmod_priv *priv, VCL_STRING name)
...
+	(void)ctx;
+	int i, ovector[VRE_MAX_GROUPS];
+	struct cookie *cookie = NULL;
+	struct cookie *current;
+	vre_t *vre = NULL;
+
+	if (expression == NULL || *expression == '\0')
+		return(NULL);
+
+	if (priv_call->priv == NULL) {
```
Comment (nigoroll) [inline src/vmod_cookie.c]: I think this would need to happen under a lock, otherwise the two compilation calls will race and we'll leak

Comment (nigoroll) [reply src/vmod_cookie.c]: An alternative would be to use the object interface and prep the RE in the object constructor

File: src/vmod_cookie.c [hunk 2]
```diff
@@ -213,6 +220,58 @@ vmod_get(VRT_CTX, struct vmod_priv *priv, VCL_STRING name)
...
 
+vre_t *
+compile_re(VRT_CTX, VCL_STRING expression) {
+	vre_t *vre;
+	const char *error;
+	int erroroffset;
+
+	vre = VRE_compile(expression, 0, &error, &erroroffset);
+	if (vre == NULL) {
+		VSLb(ctx->vsl, SLT_VCL_Log, "cookie: PCRE compile error at char %i: %s", erroroffset, error);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: SLT_Error? Or even better VCL_fail?

File: src/vmod_cookie.c [hunk 3]
```diff
@@ -213,6 +220,58 @@ vmod_get(VRT_CTX, struct vmod_priv *priv, VCL_STRING name)
...
+		if (!vre)
+			return(NULL);   // Not much else to do, error already logged.
+
+		priv_call->priv = vre;
+		priv_call->free = free;
+	}
+
+	VTAILQ_FOREACH(current, &vcp->cookielist, list) {
+		CHECK_OBJ_NOTNULL(current, VMOD_COOKIE_ENTRY_MAGIC);
+		VSLb(ctx->vsl, SLT_VCL_Log, "cookie: checking %s", current->name);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: would we really want to produce that much log traffic?
One option would be to have a debug flag in the vmod.

File: src/vmod_cookie.c [hunk 4]
```diff
@@ -213,6 +220,58 @@ vmod_get(VRT_CTX, struct vmod_priv *priv, VCL_STRING name)
...
+		i = VRE_exec(vre, current->name, strlen(current->name), 0, 0, ovector,
+				     VRE_MAX_GROUPS, NULL);
+		if (i >= 0) {
+			VSLb(ctx->vsl, SLT_VCL_Log, "cookie: %s is a match for regex '%s'", current->name, expression);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: same...

File: src/vmod_cookie.c [hunk 5]
```diff
@@ -307,18 +366,82 @@ filter_cookies(struct vmod_priv *priv, VCL_STRING list_s,
...
+VCL_VOID
+re_filter(VRT_CTX, struct vmod_priv *priv, struct vmod_priv *priv_call, VCL_STRING expression, int mode) {
+	struct vmod_cookie *vcp = cobj_get(priv);
+	struct cookie *current, *safeptr;
+	(void)ctx;
+	int i, ovector[VRE_MAX_GROUPS];
+	vre_t *vre = NULL;
+
+	if (priv_call->priv == NULL) {
+		vre = compile_re(ctx, expression);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: see above (lock)

File: src/vmod_cookie.c [hunk 6]
```diff
@@ -307,18 +366,82 @@ filter_cookies(struct vmod_priv *priv, VCL_STRING list_s,
...
+
+	VTAILQ_FOREACH_SAFE(current, &vcp->cookielist, list, safeptr) {
+		CHECK_OBJ_NOTNULL(current, VMOD_COOKIE_ENTRY_MAGIC);
+
+		i = VRE_exec(priv_call->priv, current->name, strlen(current->name),
+					 0, 0, ovector, VRE_MAX_GROUPS, NULL);
+
+		if (mode == FILTER_ACTION_BLACKLIST) {
+			if (i >= 0) {
+				VSLb(ctx->vsl, SLT_VCL_Log, "Removing matching cookie %s (value: %s)", current->name, current->value);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: logging

File: src/vmod_cookie.c [hunk 7]
```diff
@@ -307,18 +366,82 @@ filter_cookies(struct vmod_priv *priv, VCL_STRING list_s,
...
+				VTAILQ_REMOVE(&vcp->cookielist, current, list);
+			}
+		} else if (mode == FILTER_ACTION_WHITELIST) {
+			if (i >= 0) {
+				VSLb(ctx->vsl, SLT_VCL_Log, "Cookie %s matches expression '%s'", current->name, expression);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: logging

File: src/vmod_cookie.c [hunk 8]
```diff
@@ -307,18 +366,82 @@ filter_cookies(struct vmod_priv *priv, VCL_STRING list_s,
...
+			} else {
+				VSLb(ctx->vsl, SLT_VCL_Log, "Removing cookie %s (value: %s)", current->name, current->value);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: logging

**Comment by lkarsten** [comment]
@nigoroll I think I got it all. Feedback appreciated.

**Comment by lkarsten** [comment]
Hi. 

Is there anything I can do to move this along?

**Comment by marcaddeo** [comment]
Would love to see this merged in as well. Would be very useful in the case of randomized session cookie identifiers.

**Comment by nigoroll** [comment]
@lkarsten as you know I am not the maintainer, but I would love to merge this into my master fork at https://github.com/nigoroll/varnish-modules And with that hat on I would appreciate if you consolidated your PR into a smaller number of commits (one?)

File: src/vmod_cookie.c [hunk 9]
```diff
@@ -248,21 +252,24 @@ vmod_get_re(VRT_CTX, struct vmod_priv *priv, struct vmod_priv *priv_call, VCL_ST
 		return(NULL);
 
 	if (priv_call->priv == NULL) {
+		AZ(pthread_mutex_lock(&mtx));
 		vre = compile_re(ctx, expression);
 		if (!vre)
 			return(NULL);   // Not much else to do, error already logged.
```
Comment (nigoroll) [inline src/vmod_cookie.c]: missing mutex unlock

File: src/vmod_cookie.c [hunk 10]
```diff
@@ -248,21 +252,24 @@ vmod_get_re(VRT_CTX, struct vmod_priv *priv, struct vmod_priv *priv_call, VCL_ST
...
 	VTAILQ_FOREACH(current, &vcp->cookielist, list) {
 		CHECK_OBJ_NOTNULL(current, VMOD_COOKIE_ENTRY_MAGIC);
-		VSLb(ctx->vsl, SLT_VCL_Log, "cookie: checking %s", current->name);
+
+		// VSLb(ctx->vsl, SLT_VCL_Log, "cookie: checking %s", current->name);
 		i = VRE_exec(vre, current->name, strlen(current->name), 0, 0, ovector,
 				     VRE_MAX_GROUPS, NULL);
 		if (i >= 0) {
-			VSLb(ctx->vsl, SLT_VCL_Log, "cookie: %s is a match for regex '%s'", current->name, expression);
+			// VSLb(ctx->vsl, SLT_VCL_Log, "cookie: %s is a match for regex '%s'", current->name, expression);
```
Comment (nigoroll) [inline src/vmod_cookie.c]: if you really want to keep your debug logging in, maybe use a macro which is only enabled by a define?

**Comment by gquintard** [comment]
Hi @lkarsten, sorry for leaving you hanging. We'll gladly take the request once @nigoroll's remarks have been address and the commits squashed. I'll have one last review right before merging but I trust Nils to have caught pretty much everything

**Comment by lkarsten** [comment]
Hi @gquintard .
 
I've since moved to other tasks with other customers, and I don't have the time or inclination to follow up on this.

If anyone needs this, please be in touch and I can help you along.

-Lasse
 
Best of luck, and thanks to everyone that helped with review of the code.

**Comment by nigoroll** [comment]
continued in #146 after consent 1:1 from @lkarsten

**Pull request closed without merging**